### ISSUE TITLE
Inconsistent representation of the backslash character in search patterns

### ISSUE THREAD

**Issue by RedXanadu** [coreruleset/coreruleset#2179]
### Describe the bug

**In brief**: _Inconsistent representation of the backslash character in search patterns means that some CRS rules behave subtly differently depending on the platform in use._

Following on from #2140, a search of the entire Core Rule Set (`v3.4/dev` branch) has revealed a handful of instances of the pattern `\\` being used to represent a single backslash character.

The pattern `\\` works correctly with `libmodsecurity` _however_ Apache requires the pattern `\\\\` in order to correctly represent a single backslash character (due to slight differences in how rules are parsed, as explained in detail [here](https://github.com/SpiderLabs/ModSecurity/issues/2148)).

Most CRS rules use a portable representation of a backslash, e.g. `[\\\\]`, which works as intended with **both** `libmodsecurity` _and_ Apache. Some rules only use `\\`, however. This inconsistency means that some CRS rules behave subtly differently depending on the platform in use.

### Steps to reproduce

Test one of the rules highlighted below on both Apache/`mod_security2` and nginx/`libmodsecurity`.

Test using a pattern containing a `\` character that should match.

Observe that the rule may be triggered on one platform but not the other, e.g. the rule may match with nginx but not match with Apache.

---

#### Example

Attempting to trigger rule 933210 with the pattern `(sys\tem)('uname');`

Command:
`curl -o /dev/null -v "localhost:80/?test=(sys\tem)('uname');"`

**Test 1**
Testing against Apache/`mod_security2`:
`< HTTP/1.1 200 OK`
(No rules triggered)

**Test 2**
Testing against nginx/`libmodsecurity`:
`< HTTP/1.1 403 Forbidden`
```
…ModSecurity: Warning. Matched "Operator `Rx'…against variable `ARGS:test' (Value: `(sys\tem)('uname');' ) …[id "933210"]…
```

---

### Expected behaviour

All CRS rules are expected to perform similarly on either Apache/`mod_security2` or `libmodsecurity`.

### Actual behaviour

Rules using the pattern `\\` behave differently depending on the platform.

### Additional context

This is much clearer when run on a terminal with colour highlighting:

```
$ grep -E '[^\\]\\\\[^\\]' rules/*
rules/php-errors.data:Cannot access property started with '\\0'
rules/REQUEST-930-APPLICATION-ATTACK-LFI.conf:SecRule REQUEST_URI|ARGS|REQUEST_HEADERS|!REQUEST_HEADERS:Referer|XML:/* "@rx (?:(?:^|[\\/])\.\.[\\/]|[\\/]\.\.(?:[\\/]|$))" \
rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:# [\^\.\w '\"/\\\\]*\\\\)?[\"\^]*       \\net\share\dir\cmd
rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?:[*?`\\'][^/\n]+/|\$[({\[#a-zA-Z0-9]|/[^/]+?[*?`\\'])" \
rules/REQUEST-933-APPLICATION-ATTACK-PHP.conf:SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|REQUEST_FILENAME|ARGS_NAMES|ARGS|XML:/* "@rx (?:(?:\(|\[|\")[a-zA-Z0-9_.$\"'\[\](){}*\s\\]+(?:\)|\]|\")[0-9_.$\"'\[\](){}*\s]*\([a-zA-Z0-9_.$\"'\[\](){}*\s].*\)|\([\s]*string[\s]*\)[\s]*(?:\"|'))\s*[;]" \
rules/REQUEST-941-APPLICATION-ATTACK-XSS.conf:SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|REQUEST_HEADERS:User-Agent|REQUEST_HEADERS:Referer|ARGS_NAMES|ARGS|XML:/* "@rx (?i)(?:\W|^)(?:javascript:(?:[\s\S]+[=\\\(\[\.<]|[\s\S]*?(?:\bname\b|\\[ux]\d))|data:(?:(?:[a-z]\w+/\w[\w+-]+\w)?[;,]|[\s\S]*?;[\s\S]*?\b(?:base64|charset=)|[\s\S]*?,[\s\S]*?<[\s\S]*?\w[\s\S]*?>))|@\W*?i\W*?m\W*?p\W*?o\W*?r\W*?t\W*?(?:/\*[\s\S]*?)?(?:[\"']|\W*?u\W*?r\W*?l[\s\S]*?\()|\W*?-\W*?m\W*?o\W*?z\W*?-\W*?b\W*?i\W*?n\W*?d\W*?i\W*?n\W*?g[\s\S]*?:[\s\S]*?\W*?u\W*?r\W*?l[\s\S]*?\(" \
rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf:SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|!REQUEST_COOKIES:/_pk_ref/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\\x00)" \
```

The `\\net\share\dir\cmd` line is irrelevant, as it is a comment. The remaining lines appear to be valid issues (and easy fixes, I believe).

So, affected are the following:

- The `php-errors.data` file (which is referenced only by (response) rule 953100, PL1)
- Rule 930110, PL1 (already being handled in #2140)
- Rule 932200, PL2
- Rule 933210, PL1
- Rule 941170, PL1
- Rule 942440, PL2

### Your Environment

* CRS version (e.g., v3.2.0): v3.4 (dev branch)
* Paranoia level setting: Any
* ModSecurity version (e.g., 2.9.3): 2.9.4 with Apache, 3.0.5 with nginx
* Web Server and version (e.g., apache 2.4.41): Apache 2.4, nginx 1.20
* Operating System and version: Whatever flavour of Linux the `modsecurity-crs-docker` containers use

**Comment by franbuehler** [comment]
Thank you very much for looking for these patterns and for this very clearly structured issue!
If I get it right, `\\\\` would work for both platforms.
So we will have to open a PR that replaces `\\` with `\\\\`?

**Comment by fzipi** [comment]
@RedXanadu Thanks for keep looking at this :)

I had doubts with 941170. A quick look at https://github.com/hackademix/noscript/blob/8115531e36aadf07de4394ec609f99211f0f2530/src/xss/InjectionChecker.js#L855 reassures me that we should be using `\\\\`, so I'm changing that one.

I don't think we should change 942440, as the regexp there is trying to match `\x00` (null char). What do you think?

**Comment by fzipi** [comment]
Additionally, rule 953100 uses `pmFromFile` instead of pcre regex matching, so it should not be parsed with the exact same parser, and we should leave it as it.

**Comment by RedXanadu** [comment]
I've re-tested rule 942440, using the payload `' or 1=1;\x00` which should be caught (it is listed in the comments as an example payload).

---

**Test command:**
```
curl -o /dev/null -v localhost:80/post --data "' or 1=1;\x00"
```

**Test against modsec2-apache:**

Rule 942440 is not triggered, which I believe is a false negative

**Test against modsec3-nginx:**

Rule 942440 is triggered:

```
2021/08/30 20:21:03
    [info] 23#23: *1 ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\\x00)' against variable `ARGS:' or 1' (Value: `1;\x00' )
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"]
    [line "1248"]
    [id "942440"]
    [rev ""]
    [msg "SQL Comment Sequence Detected"]
    [data "Matched Data: ;\x00 found within ARGS:' or 1: 1;\x00"]
    [severity "2"]
    [ver "OWASP_CRS/3.4.0-dev"]
    [maturity "0"]
    [accuracy "0"]
    [tag "application-multi"]
    [tag "language-multi"]
    [tag "platform-multi"]
    [tag "attack-sqli"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/248/66"]
    [tag "PCI/6.5.2"]
    [tag "paranoia-level/2"]
    [hostname "172.29.0.3"]
    [uri "/post"]
    [unique_id "163035486384.657590"]
    [ref "o1,5v150,6t:urlDecodeUni"], client: 172.29.0.1, server: localhost, request: "POST /post HTTP/1.1", host: "localhost:81"
```

---

I believe that rule 942440 is an instance where it is possible and appropriate to modify the regular expression to fix this issue:

```
From
    @rx (?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\\x00)
to
    @rx (?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?[\\\\]x00)
                                                                             ~~~~~~
```

**Re-test against modsec2-apache:**

Rule 942440 is triggered:

```
[Mon Aug 30 20:37:17.632977 2021]
    [:error]
    [pid 13:tid 140558360053504]
    [client 172.29.0.1:42920]
    [client 172.29.0.1] ModSecurity: Warning. Pattern match "(?:/\\\\*!?|\\\\*/|[';]--|--[\\\\s\\\\r\\\\n\\\\v\\\\f]|--[^-]*?-|[^&-]#.*?[\\\\s\\\\r\\\\n\\\\v\\\\f]|;?[\\\\\\\\]x00)" at ARGS:' or 1.
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"]
    [line "1267"]
    [id "942440"]
    [msg "SQL Comment Sequence Detected"]
    [data "Matched Data: ;\\x5cx00 found within ARGS:' or 1: 1;\\x5cx00"]
    [severity "CRITICAL"]
    [ver "OWASP_CRS/3.4.0-dev"]
    [tag "application-multi"]
    [tag "language-multi"]
    [tag "platform-multi"]
    [tag "attack-sqli"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/248/66"]
    [tag "PCI/6.5.2"]
    [tag "paranoia-level/2"]
    [hostname "localhost"]
    [uri "/post"]
    [unique_id "YS1BfWuEYcXvRcyOp4MGGgAAAEA"]
```

**Re-test against modsec3-nginx:**

Rule 942440 is triggered:

```
2021/08/30 20:38:14
    [info] 24#24: *1 ModSecurity: Warning. Matched "Operator `Rx' with parameter `(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?[\\\\]x00)' against variable `ARGS:' or 1' (Value: `1;\x00' )
    [file "/etc/modsecurity.d/owasp-crs/rules/REQUEST-942-APPLICATION-ATTACK-SQLI.conf"]
    [line "1248"]
    [id "942440"]
    [rev ""]
    [msg "SQL Comment Sequence Detected"]
    [data "Matched Data: ;\x00 found within ARGS:' or 1: 1;\x00"]
    [severity "2"]
    [ver "OWASP_CRS/3.4.0-dev"]
    [maturity "0"]
    [accuracy "0"]
    [tag "application-multi"]
    [tag "language-multi"]
    [tag "platform-multi"]
    [tag "attack-sqli"]
    [tag "OWASP_CRS"]
    [tag "capec/1000/152/248/66"]
    [tag "PCI/6.5.2"]
    [tag "paranoia-level/2"]
    [hostname "172.29.0.3"]
    [uri "/post"]
    [unique_id "163035589470.878751"]
    [ref "o1,5v150,6t:urlDecodeUni"], client: 172.29.0.1, server: localhost, request: "POST /post HTTP/1.1", host: "localhost:81"
```

---

**In summary:**

I believe this is a case where using the `[\\\\]` workaround is appropriate. Does anyone else agree/disagree? :)

**Comment by fzipi** [comment]
Even if I disagree, nothing beats good tests. I think you are right, and we should fix this one.

**Comment by RedXanadu** [comment]
@fzipi Thinking about it some more… do you think rule 942440 is designed to look for a _literal_ null character, or is it looking for the string `\x00`?
I think that makes a difference

**Comment by fzipi** [comment]
Should be looking at NUL, right?

**Comment by fzipi** [comment]
Looking at history, this might be the reference: https://www.trustwave.com/en-us/resources/blogs/spiderlabs-blog/modsecurity-sql-injection-challenge-lessons-learned/

**Comment by RedXanadu** [comment]
Okay, now I understand what you mean :)

I agree with you: adding `[\\\\]` is incorrect. I originally thought the rule was looking for the string representation `\x00` and not a real NUL byte.

---

If the rule is looking for a NUL byte, should the end of the pattern be: `…|;?\x00)` ?

I've tested how the pattern from rule 942440 is interpreted using the tools `pcre4msc2` and `pcre4msc3`, and because of the differences with escaping the result is different:

```
$ echo | ./src/pcre4msc2 -d regexes/942440_1.txt 

RAW pattern:
============
(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\\x00)

ESCAPED pattern:
================
(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\x00)
```

```
$ echo | ./src/pcre4msc3 -d regexes/942440_1.txt 

PATTERN:
========
(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\\x00)
```

ModSecurity v3 leaves the pattern with the extra `\`.

---

If the pattern is modified to `…|;?\x00)` then it ends up being the same:

```
$ echo | ./src/pcre4msc2 -d 942440_1_modified.txt 

RAW pattern:
============
(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\x00)

ESCAPED pattern:
================
(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\x00)
```

```
$ echo | ./src/pcre4msc3 -d 942440_1_modified.txt 

PATTERN:
========
(?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\x00)
```

So, that modified pattern would work the same on both platforms, I think.

I'm not sure how to test it for real: I've struggled to submit a test request containing a NUL byte using `cURL`‥

**Comment by fzipi** [comment]
You can always send a null byte encoded as `%00` I guess...

**Comment by dune73** [comment]
Awesome writeup and discussion.

**Comment by RedXanadu** [comment]
> You can always send a null byte encoded as `%00` I guess...

@fzipi Ah-hah, yes! The rule includes `t:urlDecodeUni`, so that makes a lot of sense :)

Using that to test, I can confirm that there is a difference.

---

The `modsec3-nginx` container matches on the following string (no actual null character, just the characters `\x00`):

Submitted payload:
`' or 1=1;\x00`

```
[data "Matched Data: ;\x00 found within ARGS:a: \' or 1=1;\x00"]
```

I believe this is incorrect and _should not match_.

The `modsec2-apache` container does not match, which I believe is correct.

---

The `modsec2-apache` container matches on a real null character, which is submitted using URL encoding:

Submitted payload:
`' or 1=1;%00`

```
[data "Matched Data: ;\\x00 found within ARGS:a: \\x5c' or 1=1;\\x00"]
```

I believe this is correct.

The `modsec3-nginx` container does not match, which I believe is a false negative.

---

When the regular expression is changed like so:

```
From
    @rx (?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\\x00)
to
    @rx (?:/\*!?|\*/|[';]--|--[\s\r\n\v\f]|--[^-]*?-|[^&-]#.*?[\s\r\n\v\f]|;?\x00)
```

then **both** containers match against the payload with the URL encoded null byte:

`' or 1=1;%00`

and _neither_ container matches against the string:

`' or 1=1;\x00`

I believe this is an appropriate fix, as the rule is looking for a null character and **not** simply the string "\x00". What do you think?

**Comment by fzipi** [comment]
First, I just can't stress enough how good your tests have been. This is really awesome.

I agree that it is exactly what I would expect from this, and that this fixes the behavior as intended.

**Comment by RedXanadu** [comment]
@fzipi Thank you :)

I've opened #2201 which I believe addresses the regex pattern issue discussed above

**Comment by RedXanadu** [comment]
Have added a new regression test to the PR: test that the input `' or 1=1;%00` is correctly caught

**Comment by dune73** [comment]
Issue covered in the [September issue chat](https://github.com/coreruleset/coreruleset/issues/2185#issuecomment-913804183).

The issue should be solved via PRs #2183 and #2201. We're keeping the issue open while the PRs are being tested.
